cli/WFCORE-5500_Evolved_Control_Flow.adoc

```diff
@@ -0,0 +1,284 @@
+and writing CLI scripts to adapt to the server configuration is required.
+
+For example, it is a common practice to check for the existence of multiple subsystems, to iterate array of resources and 
+take decisions according to the configuration (or env variables) state. Current CLI support handles such cases in a very limited way.
+
+We are here not defining a new programming language (that is really out of scope of the CLI command line utility) but we 
+are evolving existing commands to solve the major blocking points that forbid its usage in complex scenario.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFCORE-5500[WFCORE-5500]
+
+=== Related Issues
+
+* Completion issue with control flow commands: https://issues.redhat.com/browse/WFCORE-3489[WFCORE-3489]
+
+* Variables handling: https://issues.redhat.com/browse/WFCORE-5388[WFCORE-5388]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* TBD
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* https://github.com/wildfly/wildfly-core[WildFly core, CLI component]
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Nested workflow commands
+
+* We are removing the limitation to a single level of nesting. 
+* ```if``` can be nested without limitations inside any workflow command blocks. 
+* ```for``` can be nested without limitations inside any workflow command blocks. 
+* ```try``` can be nested without limitations inside any workflow command blocks.
+* Command completion takes into account nesting and proposes the right associated operations 
+(e.g.: ```catch```, ```finally```, ```end-try``` for the ```try``` command).
+* ```batch``` commands completion takes into account nesting.
+
+For example:
+
+```
+if outcome==success of /subsystem=security:read-resource
+  if outcome==success of /subsystem=elytron:read-resource
+    echo We have legacy security and elytron
+  else
+     for domain in /subsystem=security:read-children-names(child-type=security-domain)
+        echo $domain
+        for elytronDomain in /subsystem=elytron:read-children-names(child-type=security-domain)
+          batch
+            ...
+          run-batch
+        done
+     done
+  end-if
+end-if
+```
+
+=== ```if``` command new operators for Array.
+
+* Add the ability to test the size of an array by using the operator ```#=```
+* Add the ability to test if an array contains a String element by using the operator ```*=```
```

Comment:
how much are we limited syntactically? Could we name that operator `contains`?

```
if (result contains "ManagementDomain") of /subsystem=elytron:read-child-names(child-type=security-domain)
   /subsystem=elytron/security-domain=ManagementDomain:remove()
 end-if 
```

Reply:
That is a limitation, we need special character for the operator.